pkg/virtctl/create/vm.go

```diff
@@ -0,0 +1,321 @@
+	name                   string
+	instancetype           string
+	preference             string
+	imageSource            string
+	runStrategy            string
+	terminationGracePeriod int64
+}
+
+type optionFn func(*createVM, *v1.VirtualMachine) error
+
+var optFns = map[string]optionFn{
+	InstancetypeFlag:      withInstancetype,
+	PreferenceFlag:        withPreference,
+	ImageSourceFlag:       withImageSource,
+	DirectImageSourceFlag: withDirectImageSource,
+}
+
+func NewVirtualMachineCommand() *cobra.Command {
+	c := defaultCreateVM()
+	cmd := &cobra.Command{
+		Use:     VM,
+		Short:   "Create a VirtualMachine manifest.",
+		Example: c.usage(),
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			return c.run(cmd)
+		},
+	}
+
+	cmd.Flags().StringVar(&c.name, NameFlag, c.name, "Specify the name of the VirtualMachine.")
+
+	cmd.Flags().StringVar(&c.instancetype, InstancetypeFlag, c.instancetype, "Specify the Instancetype of the VirtualMachine.")
+	cmd.Flags().StringVar(&c.preference, PreferenceFlag, c.preference, "Specify the Preference of the VirtualMachine.")
+
+	cmd.Flags().StringVar(&c.imageSource, ImageSourceFlag, c.imageSource, "Specify the image source to be cloned by the VirtualMachine.")
+	cmd.Flags().StringVar(&c.imageSource, DirectImageSourceFlag, c.imageSource, "Specify the image source to be used by the VirtualMachine.")
+	cmd.MarkFlagsMutuallyExclusive(ImageSourceFlag, DirectImageSourceFlag)
+
+	cmd.Flags().StringVar(&c.runStrategy, RunStrategyFlag, c.runStrategy, "Specify the RunStrategy of the VirtualMachine.")
+	cmd.Flags().Int64Var(&c.terminationGracePeriod, TerminationGracePeriodFlag, c.terminationGracePeriod, "Specify the termination grace period of the VirtualMachine.")
+
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+
+	return cmd
+}
+
+func defaultCreateVM() createVM {
+	return createVM{
+		name:                   "vm-" + rand.String(5),
+		runStrategy:            string(v1.RunStrategyAlways),
+		terminationGracePeriod: 180,
+	}
+}
+
+func (c *createVM) run(cmd *cobra.Command) error {
+	vm := newVM(c)
+	for flag, optFn := range optFns {
+		if cmd.Flags().Changed(flag) {
+			if err := optFn(c, vm); err != nil {
+				return err
+			}
+		}
+	}
+
+	out, err := yaml.Marshal(vm)
+	if err != nil {
+		return err
+	}
+
+	cmd.Print(string(out))
+	return nil
+}
+
+func (c *createVM) usage() string {
+	return `  # Create a VirtualMachine with a random name:
+  {{ProgramName}} create vm
+    
+  # Create a VirtualMachine with a specified name
+  {{ProgramName}} create vm --name=my-vm
+  
+  # Create a VirtualMachine with a specified Instancetype
+  {{ProgramName}} create vm --instancetype=my-instancetype
+  
+  # Create a VirtualMachine with a specified Preference
+  {{ProgramName}} create vm --preference=my-preference
+  
+  # Create a VirtualMachine with a specified Instancetype, Preference and image source
+  {{ProgramName}} create vm --instancetype=my-instancetype --preference=my-preference --image-source=pvc/my-pvc
```

Comment:
What about breaking this down into individual `Volume` switches for each type of `Volume` chain we could support here?

```
--volume-pvc namespace/name       // VolumeSource -> PersistentVolumeClaim -> PersistentVolumeClaimVolumeSource
--volume-dv namespace/name        // VolumeSource -> DataVolumeSource
--volume-dvt-pvc namespace/name   // VolumeSource -> DataVolumeSource -> DataVolumeTemplate -> DataVolumeSourcePVC
--volume-dvt-ds  namespace/name   // VolumeSource -> DataVolumeSource -> DataVolumeTemplate -> DataVolumeSourceRef -> DataSource
```

We could easily have duplicated variants of these with fully expanded resource kinds, the above is just easier short hand.

Reply:
I like this idea, also leaves it open to implement volume populators with a similar syntax in the future.